visionmetrics/key_value_pair/key_value_pair_eval_base.py

```diff
@@ -0,0 +1,160 @@
+from enum import Enum
+from functools import reduce
+import logging
+import operator
+import torch
+from torchmetrics import Metric
+
+# Import relevant visionmetrics modules; even though they appear to be unused to flake8, they are needed for dynamic metric instantiation at runtime.
+from visionmetrics import caption, classification, detection, grounding, regression  # noqa: F401
+
+logger = logging.getLogger(__name__)
+
+
+class SupportedKeyWiseMetric(str, Enum):
+    Caption_AzureOpenAITextModelCategoricalScore = "caption.AzureOpenAITextModelCategoricalScore"
+    Classification_MulticlassAccuracy = "classification.MulticlassAccuracy"
+    Classification_MultilabelAccuracy = "classification.MultilabelAccuracy"
+    Classification_MulticlassF1 = "classification.MulticlassF1Score"
+    Classification_MultilabelF1 = "classification.MultilabelF1Score"
+    Detection_MeanAveragePrecision = "detection.MeanAveragePrecision"
+    Regression_MeanAbsoluteError = "regression.MeanAbsoluteError"
+
+
+class KeyValuePairEvaluatorBase(Metric):
+    """
+    Evaluator for evaluating key-value pair datasets, where each key in the labels represents an extracted field with a consistent type (string, number, etc.).
+    It accepts a dictionary mapping the keys to corresponding metric information, and returns the corresponding key-wise set of metrics.
+    Each key can have a different metric for evaluation. The metrics supported are specified in SupportedKeyWiseMetric.
+
+    Args:
+        key_metric_map: dictionary from keys (extracted field names) to a dictionary with four required fields:
+        1. metric_name: string of the metric name as defined in visionmetrics (e.g., 'classification.MulticlassAccuracy'), which should be among the ones specified in SupportedKeyWiseMetric.
+        2. metric_args: dictionary of args to pass in as keyword arguments to the initialization function of the metric.
+        3. preprocessor: function object that can be called with (prediction, target) values for a single instance to preprocess them into the desired format for the corresponding metric.
```

Comment:
do we expect preprocessor to required coupled gt and pred? Otherwise, can we separate to gt_preprocessor and pred_preprocessor?  In some cases they are same or one can be adapted from another

Reply:
Good idea to decouple them, thanks!